genai-perf/docs/goodput_tutorial.md

```diff
@@ -0,0 +1,558 @@
+```
+
+Example output:
+
+```
+                                   LLM Metrics
+┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
+┃                Statistic ┃    avg ┃    min ┃    max ┃    p99 ┃    p90 ┃    p75 ┃
+┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
+│ Time to first token (ms) │   5.50 │   3.78 │  19.25 │  19.20 │   5.86 │   4.98 │
+│ Inter token latency (ms) │   1.52 │   1.35 │   1.67 │   1.66 │   1.62 │   1.58 │
+│     Request latency (ms) │ 176.02 │ 165.97 │ 197.03 │ 196.99 │ 181.80 │ 177.86 │
+│   Output sequence length │ 113.34 │ 107.00 │ 127.00 │ 125.14 │ 118.00 │ 116.00 │
+│    Input sequence length │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │
+└──────────────────────────┴────────┴────────┴────────┴────────┴────────┴────────┘
+Output token throughput (per sec): 1287.08
+Request throughput (per sec): 11.36
+Request goodput (per sec): 10.65
+```
+
+### Run GenAI-Perf with invalid goodput constraints
+
+```bash
+genai-perf profile \
+  -m gpt2 \
+  --service-kind triton \
+  --backend vllm \
+  --num-prompts 100 \
+  --random-seed 123 \
+  --synthetic-input-tokens-mean 200 \
+  --synthetic-input-tokens-stddev 0 \
+  --streaming \
+  --output-tokens-mean 100 \
+  --output-tokens-stddev 0 \
+  --output-tokens-mean-deterministic \
+  --tokenizer hf-internal-testing/llama-tokenizer \
+  --concurrency 2 \
+  --measurement-interval 800 \
+  --profile-export-file my_profile_export.json \
+  --url localhost:8001 \
+  --goodput time_to_first_tokens:8 inter_token_latencies:2 output_token_throughputs_per_requesdt:650
+```
+
+Example output:
+
+```
+2024-08-14 15:18 [INFO] genai_perf.goodput_calculator.llm_goodput_calculator:92 - Invalid SLOs found: time_to_first_tokens, inter_token_latencies, output_token_throughputs_per_requesdt. The goodput will be -1. Valid SLOs are: time_to_first_token, inter_token_latency, request_latency, output_token_throughput_per_request.
+                                   LLM Metrics
+┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
+┃                Statistic ┃    avg ┃    min ┃    max ┃    p99 ┃    p90 ┃    p75 ┃
+┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
+│ Time to first token (ms) │   5.19 │   3.88 │  16.79 │  16.77 │   5.94 │   4.96 │
+│ Inter token latency (ms) │   1.56 │   1.44 │   1.69 │   1.68 │   1.63 │   1.60 │
+│     Request latency (ms) │ 176.93 │ 170.95 │ 197.70 │ 197.68 │ 177.86 │ 176.21 │
+│   Output sequence length │ 111.47 │ 103.00 │ 123.00 │ 122.38 │ 118.80 │ 114.00 │
+│    Input sequence length │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │
+└──────────────────────────┴────────┴────────┴────────┴────────┴────────┴────────┘
+Output token throughput (per sec): 1259.65
+Request throughput (per sec): 11.30
+Request goodput (per sec): -1.00
+```
+
+### Run GenAI-Perf with no goodput constraints
```

Comment:
We don't need to show this either.

Reply:
Updated the tutorial, skip replying the rest of this file.